articles/dms/tutorial-mysql-azure-mysql-offline-portal.md

```diff
@@ -0,0 +1,254 @@
+* Open your Windows firewall to allow connections from Virtual Network for Azure Database Migration Service to access the source MySQL Server, which by default is TCP port 3306.
+* When using a firewall appliance in front of your source database(s), you may need to add firewall rules to allow connections from Virtual Network for Azure Database Migration Service to access the source database(s) for migration.
+* Create a server-level [firewall rule](../azure-sql/database/firewall-configure.md) or [configure VNET service endpoints](../mysql/howto-manage-vnet-using-portal.md) for target Azure Database for MySQL to allow Virtual Network for Azure Database Migration Service access to the target databases.
+* The source MySQL must be on supported MySQL community edition. To determine the version of MySQL instance, in the MySQL utility or MySQL Workbench, run the following command:
+
+    ```
+    SELECT @@version;
+    ```
+
+* Azure Database for MySQL supports only InnoDB tables. To convert MyISAM tables to InnoDB, see the article [Converting Tables from MyISAM to InnoDB](https://dev.mysql.com/doc/refman/5.7/en/converting-tables-to-innodb.html)
+* The user must have the privileges to read data on the source database.
+
+## Migrate database schema
+
+To transfer all the database objects like table schemas, indexes and stored procedures, we need to extract schema from the source database and apply to the target database. To extract schema, you can use mysqldump with the `--no-data` parameter. For this you need a machine which can connect to both the source MySQL database and the target Azure Database for MySQL.
+
+To export the schema using mysqldump, run the following command:
+
+```
+mysqldump -h [servername] -u [username] -p[password] --databases [db name] --no-data > [schema file path]
+```
+
+For example:
+
+```
+mysqldump -h 10.10.123.123 -u root -p --databases migtestdb --no-data > d:\migtestdb.sql
+```
+
+To import schema to target Azure Database for MySQL, run the following command:
+
+```
+mysql.exe -h [servername] -u [username] -p[password] [database]< [schema file path]
+ ```
+
+For example:
+
+```
+mysql.exe -h mysqlsstrgt.mysql.database.azure.com -u docadmin@mysqlsstrgt -p migtestdb < d:\migtestdb.sql
+ ```
+
+If you have foreign keys in your schema, the parallel data load during migration will be handled by the migration task. There is no need to drop foreign keys during schema migration.
+
+If you have triggers in the database (insert or update triggers), it will enforce data integrity in the target ahead of full data migration from the source. The recommendation is to disable triggers on all the tables in the target during migration, and then enable the triggers after migration is done.
+
+Execute the following script in MySQL Workbench on the target database to extract the drop trigger script and add trigger script.
+
+```sql
+SELECT
```

Comment:
I think there are some problems with these queries. When I ran them against one of my DB with triggers it didnt pick up all of the triggers in one of the databases. Also the database/table/trigger names are not escaped so schemas with special characters could cause problems. This also does not handled ordered triggers that use the "PRECEDES" keyword in the trigger creation. These ordered triggers need to be created in order of their "ACTION_ORDER" column in the information_schema.triggers table. More details: https://dev.mysql.com/doc/refman/8.0/en/information-schema-triggers-table.html

Reply:
Fixed
Names are escaped
Order by put for right sequence
PRECEDES is not needed as the triggers are already created in the right order. PRECEDES  is needed when you want to control the sequence because trigger with lower action order is created later.